explorer/interpreter/resolve_names.cpp

```diff
@@ -273,15 +273,28 @@ auto NameResolver::ResolveNames(Expression& expression,
       break;
     }
     case ExpressionKind::SimpleMemberAccessExpression: {
+      // If the left-hand side of the `.` is a namespace or alias to namespace,
+      // resolve the name.
       auto& access = cast<SimpleMemberAccessExpression>(expression);
       CARBON_ASSIGN_OR_RETURN(std::optional<ValueNodeView> scope,
                               ResolveNames(access.object(), enclosing_scope));
+      if (!scope) {
+        break;
+      }
 
-      // If the left-hand side of the `.` is a namespace, resolve the name.
-      // TODO: Look through aliases.
-      if (scope && isa<NamespaceDeclaration>(scope->base())) {
-        auto ns_it =
-            namespace_scopes_.find(&cast<NamespaceDeclaration>(scope->base()));
+      std::optional<Nonnull<const NamespaceDeclaration*>> namespace_decl;
+      if (isa<AliasDeclaration>(scope->base())) {
+        const AliasDeclaration& alias = cast<AliasDeclaration>(scope->base());
+        auto resolved_decl = alias.resolved_declaration();
+        if (resolved_decl && isa<NamespaceDeclaration>(resolved_decl.value())) {
+          namespace_decl = cast<NamespaceDeclaration>(resolved_decl.value());
+        }
+      } else if (isa<NamespaceDeclaration>(scope->base())) {
+        namespace_decl = &cast<NamespaceDeclaration>(scope->base());
+      }
```

Comment:
```suggestion
      if (const auto* base_namespace =
              dyn_cast<NamespaceDeclaration>(&scope->base())) {
        namespace_decl = base_namespace;
      } else if (const auto* alias = dyn_cast<AliasDeclaration>(&scope->base());
                 alias && alias->resolved_node()) {
        if (const auto* alias_namespace =
                dyn_cast<NamespaceDeclaration>(*alias->resolved_node())) {
          namespace_decl = alias_namespace;
        }
      }
```

Let's make more use of dyn_cast here instead of isa+cast